### ISSUE TITLE
samsung cups drivers fail to build due tu cups/gcc package changes (multiple outputs)

### ISSUE THREAD

**Issue by spinus** [NixOS/nixpkgs#16238]
## Issue description

In commit 5787ac63649e4e3f11b70f44fa66c7b304956c86 multiple outputs were introduced to cups.

Because output "out" is not "default" one, driver cannot be build anymore. I'm not sure why but switching the order of outputs in cups breaks cups build. I tried to fix driver package and I used `{cups.out}` instead of `{cups}`, this helped partially.

Looks like the same happened to some system libraries and samsung driver cannot find "libstdc++" anymore.

```
++ echo 'Patching ELF file: x86_64/smfpnetdiscovery'
Patching ELF file: x86_64/smfpnetdiscovery
++ patchelf --set-interpreter /nix/store/68gvzfvm4886b4ni6wfl1j5n9di69zid-glibc-2.23/lib/ld-linux-x86-64.so.2 --set-rpath /nix/store/zpg6x8x2a840nzrl3cq5mly1l7cvvkjy-cups-2.1.3/lib:/nix/store/bxn5llxrp47cxws5zj20nrwig28g6qw6-gcc-5.3.0/lib:/nix/store/6mmslibwczk3y4sdnxv5549pz9l1jcqq-gcc-wrapper-5.3.0/lib:/nix/store/65sxr06rmq3i96lmd8pclvc1jzyh1gc8-glibc-2.23-dev/lib x86_64/smfpnetdiscovery
++ ldd x86_64/smfpnetdiscovery
++ grep 'not found'
    libstdc++.so.6 => not found
++ exit 1
```

Original build command:

```
  32     my_patchelf \                                                               
  33       --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \             
  34       --set-rpath ${cups}/lib:$(cat $NIX_CC/nix-support/orig-cc)/lib:${stdenv.glibc}/lib \
  35       - ${arch}/{pstosecps,rastertospl,smfpnetdiscovery}
```

I tried to fix it like (but without any luck):

```
    31     my_patchelf \                                                               
    32       --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \             
!   33       --set-rpath ${cups.out}/lib:$(cat $NIX_CC/nix-support/orig-cc)/lib:${stdenv.cc}/lib:${stdenv.glibc}/lib \
    34       - ${arch}/{pstosecps,rastertospl,smfpnetdiscovery}
```
1. Should "out" be used as first output for backward compatibility?
2. How to include libstdc++ properly?
### Steps to reproduce

nix-env -iA nixos.samsung-unified-linux-driver_1_00_37
## Technical details
- System:16.09pre84674.fbde3a7 
- Nix version: 1.11.2
- Nixpkgs version: 16.09pre85363.886c03a

**Comment by jokogr** [comment]
I also had this issue and almost completely rewrote the nix expression. Anyway, the scanner part is still pending by my side, so I will not do a PR in the near future. Perhaps @nckx could update the package.

Try the following:

```
ln -s ${stdenv.cc.cc.lib}/lib/libstdc++.so.6 $out/lib/
```

**Comment by spinus** [comment]
For me scanner currently is not working as well.
Thanks for sharing the solution, I'll try

**Comment by nckx** [comment]
@spinus: feel free to @notify the `maintainers` when you've got a package-specific issue like this. It might go unnoticed otherwise. Unfortunately I don't currently have access to any Nix/NixOS machines *or* my Samsung printer, let alone a scanner. I'll see what I can do.

@jokogr: any WIP you could point me to? I'm curious :-)

**Comment by jokogr** [comment]
@nckx: Sure thing, have a look at https://github.com/jokogr/nixpkgs/commit/c41e771a388dca6cb5f15edab2fd36fd6f96c8a8

* Mostly I try to clean up `my_patchelf`
* I add a symlink for `libstdc++.so.6`
* I add a symlink for `gs` (ghostscript) (@spinus: you might want to add this, as well)

With this setup my Samsung M2875FD network printer works properly and `smfpnetdiscovery` detects it as a scanner, too. I still have to check sane, though, it's in my backlog.

**Comment by kierdavis** [comment]
I'm also experiencing this issue. Are we able to merge what we've got so far, even though scanner support is broken? Currently he package is uninstallable in its

**Comment by nckx** [comment]
I lost track of this a bit, sorry. I don't have access to any working Samsung printers nowadays (I even asked some friends who used to have one — no luck).

@kierdavis: I'll promptly merge anything that builds and improves the situation for those who do. Thanks!

**Comment by nckx** [comment]
@kierdavis: just FYI, people @-mentioned in subsequent edits to post won't receive mail (unless, like me, they were already subscribed).

**Comment by kierdavis** [comment]
Cool, I'll see if I can get my M2675FN working using this driver in the next few days.

**Comment by kierdavis** [comment]
Printing worked first time on my M2675FN. Haven't tested anything relating to the scanner yet. Will do a PR shortly.

**Comment by jokogr** [comment]
@kierdavis and the rest, feel free to test my PR at https://github.com/NixOS/nixpkgs/pull/20060, I have tested it with a networked M2875FD.

**Comment by nckx** [comment]
Thanks, @jokogr!

I'll go ahead and close this; if certain printers fail to print (not uncommon after Samsung driver updates), that's a separate issue.

**Comment by kierdavis** [comment]
@jokogr your PR worked for me, brilliant.

@nckx I wouldn't advise closing this just yet, as the original issue (the samsung-unified-linux-driver_1_00_37 package failing to build) hasn't been addressed. I've sent a [PR](https://github.com/NixOS/nixpkgs/pull/20092) with the contents of https://github.com/jokogr/nixpkgs/commit/c41e771a388dca6cb5f15edab2fd36fd6f96c8a8, which does build correctly and printing appears to work on M2875FD and M2675FN devices.

**Comment by nckx** [comment]
@kierdavis: yup, I'll merge that when Travis finishes.

**Comment by nckx** [comment]
(@jokogr: if that's all right with you, of course.)

**Comment by spinus** [comment]
Just tested and works on M2070 (pretty basic model). 

Thank you very much for help @kierdavis, @jokogr, @nckx !
